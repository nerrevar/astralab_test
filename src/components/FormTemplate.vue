<template>
  <div class="form">
    <FormGroup
      v-for="(item, index) in formFields"
      :key="index"
      :item="item"
      :errorFields="errorFields"
      @invalidate="invalidate($event)"
      @validate="validate($event)"
    />
    <div
      class="form__submit"
      @click="validateForm"
    >
      {{ submitText }}
    </div>
  </div>
</template>

<script>
import FormGroup from '@/components/FormGroup.vue'

export default {
  name: 'FormTemplate',
  components: { FormGroup },
  props: {
    formFields: {
      type: Array,
      required: true,
    },
    submitText: {
      type: String,
      required: true,
    },
  },
  data () {
    return { errorFields: [] }
  },
  methods: {
    invalidate (name) {
      if (this.errorFields.indexOf(name) === -1)
        this.errorFields.push(name)
    },
    validate (name) {
      const index = this.errorFields.indexOf(name)
      if (index !== -1)
        this.errorFields.splice(index, 1)
    },
    validateForm () {
      if (this.errorFields.length === 0) {
        const result = {}
        for (const child of this.$children)
          if (child.isValid())
            Object.assign(result, child.getData())

        if (Object.keys(result).length === this.formFields.length)
          this.$emit('submit', result)
      }
    },
  },
}
</script>

<style lang="sass" scoped>
.form
  border-radius: 40px
  padding: 22px
  background-color: white

  &__submit
    +ellipse
    background-color: #1e1a3e
    color: white
    text-align: center
</style>
