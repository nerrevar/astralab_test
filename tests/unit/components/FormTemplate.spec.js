import { shallowMount } from '@vue/test-utils'

import FormTemplate from '@/components/FormTemplate.vue'

describe('FormTemplate.vue', () => {
  const vm = shallowMount(FormTemplate, {
    propsData: {
      formFields: Array(1),
      submitText: 'Sign In',
    },
  })

  it('renders', () => {
    expect(vm.element).toMatchSnapshot()
  })

  it('method invalidate', () => {
    vm.vm.invalidate('name')
    expect(vm.vm.errorFields).toStrictEqual(['name'])
    vm.vm.invalidate('name')
    expect(vm.vm.errorFields).toStrictEqual(['name'])
  })

  it('method invalidate', () => {
    vm.vm.validate('name')
    expect(vm.vm.errorFields).toStrictEqual([])
    vm.vm.validate('name')
    expect(vm.vm.errorFields).toStrictEqual([])
  })

  it('method validateForm on invalid', () => {
    vm.vm.$children[0].isValid = () => false
    vm.vm.validateForm()
    expect(vm.emitted('submit')).toBeFalsy()
  })

  it('method validateForm on valid', () => {
    vm.vm.$children[0].isValid = () => true
    vm.vm.$children[0].getData = () => ({ name: 'test' })
    vm.vm.validateForm()
    expect(vm.emitted('submit')).toBeTruthy()
    expect(vm.emitted('submit')[0]).toStrictEqual([{ name: 'test' }])
  })
})
