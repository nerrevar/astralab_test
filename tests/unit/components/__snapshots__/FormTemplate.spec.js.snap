// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTemplate.vue renders 1`] = `
<div
  class="form"
>
  <formgroup-stub
    errorfields=""
  />
   
  <div
    class="form__submit"
  >
    
    Sign In
  
  </div>
</div>
`;
